File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
private:
    int m;
    int n;

    void dfs(int r, int c, vector<vector<char>>& grid) {
        if (r < 0 || r >= m || c < 0 || c >= n || grid[r][c] == '0') return;
        grid[r][c] = '0';
        dfs(r + 1, c, grid);
        dfs(r - 1, c, grid);
        dfs(r, c + 1, grid);
        dfs(r, c - 1, grid); 
    }

public:
    int numIslands(vector<vector<char>>& grid) {
        int res = 0;

        m = grid.size();
        n = grid[0].size();

        for (int r = 0; r < m; r++) {
            for (int c = 0; c < n; c++) {
                if (grid[r][c] == '1') {
                    res++;
                    dfs(r, c, grid);
                }
            }
        }

        return res;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        res = 0
        m, n = len(grid), len(grid[0])

        def dfs(r, c):
            if r < 0 or r >= m or c < 0 or c >= n or grid[r][c] == '0':
                return False
            
            grid[r][c] = '0'
            dfs(r + 1, c)
            dfs(r - 1, c)
            dfs(r, c + 1)
            dfs(r, c - 1)

            return True

        for r in range(m):
            for c in range(n):
                if dfs(r, c):
                    res += 1
        
        return res